### ISSUE TITLE
[1.18.2: see additional context] "Power Enriched by the Heart" Advancement applies incorrectly

### ISSUE THREAD

**Issue by reetamb** [infernalexp/Mining-Master#31]
### Describe the Bug

"Power Enriched by the Heart" can sometimes be granted without meeting the necessary conditions (having Heartfelt on all four pieces of armor).

### Reproduction Steps

1. Enchanted leggings and boots both with Heartfelt 2 beforehand.
2. Took both off, causing the four additional hearts to be lost.
3. Enchanted boots with Runner 3 (had Runner 2) using a Kinetic Opal
4. Closed Smithing Table GUI.
5. The advancements "Sonic Speed!" and "Power Enriched by the Heart" were granted at the same time (the former as expected).

### Expected Result

Only "Sonic Speed!" should have been granted, because it was the only one at that moment whose conditions were fulfilled.

### Operating System

Windows 11

### Mod Version

2.1.0

### Minecraft Version

1.16.5

### Forge Version

40.1.69

### Other Mods

Apotheosis affects the Enchanting Table, the leggings had Soulbound from Ensorcellation, and Enlightened End adds Smithing Table recipes.

### Additional Context

IMPORTANT: There either aren't dropdown options for MC Version 1.18.2, Mod Version 3.0.5, or I'm somehow unable to access them. The versions for this report are 1.18.2 3.0.5, NOT 1.16.5 2.1.0 (which I selected as a placeholder).

My guess is this was caused by having two armor pieces enchanted with Heartfelt 2, and when they were taken off, the lost hearts were tallied as four different pieces with Heartfelt 1.

**Comment by Dariensg** [comment]
I'm gonna be honest, I have no idea what's happening here. I've been able to replicate it both in 1.18.2 and 1.19.2. It would seem that the thing that triggers the issue is using the smithing table. I've checked and the requirements for the achievement are being stored correctly, but for some reason using the Smithing Table makes the game think that the requirements have been met. I'm going to keep looking into it, but it may take a bit.
